src/components/pagination/index.tsx

```diff
@@ -0,0 +1,262 @@
+		</div>
+	)
+}
+
+const generateNavItems = ({ currentPage, totalPages, type }) => {
+	const rawItems = Array.from({ length: totalPages }, (_, i) => i + 1)
+
+	if (type === 'numbered') {
+		return rawItems
+	}
+
+	if (type === 'truncated') {
+		return generateTruncatedList(rawItems, currentPage)
+	}
+
+	return []
+}
+
+const Nav = ({ type = 'compact' }: NavProps) => {
+	const { page: currentPage, totalPages } = usePagination()
+	const items = useMemo(
+		() => generateNavItems({ currentPage, totalPages, type }),
+		[currentPage, totalPages, type]
+	)
+
+	return (
+		<nav className={s.nav} aria-label="Pagination navigation">
+			<ButtonArrow type={type} direction="prev" />
+			{type !== 'compact' && (
+				<ul className={s['page-list']}>
+					{items.map((e, i) => {
+						if (e === 'ellipsis') {
+							// This key will be different on every re-render which is OK
+							const key = `${e}-${i}`
+							return (
+								<li className={classNames(s['page-item'])} key={key}>
+									<div className={s.ellipsis}>...</div>
+								</li>
+							)
+						}
+
+						const isSelected = e === currentPage
+						return (
+							<li
+								className={classNames(s['page-item'], {
+									[s['is-selected']]: isSelected,
+								})}
+								// This key will be stable between re-renders
+								// which prevents react from removing focus from
+								// the previous `document.activeElement`
+								key={e}
+							>
+								<ButtonNumber page={e} />
+							</li>
+						)
+					})}
+				</ul>
+			)}
+			<ButtonArrow type={type} direction="next" />
+		</nav>
+	)
+}
+
+const ButtonArrow = ({ type, direction }: ButtonArrowProps) => {
+	const pagination = usePagination()
+	const { handleClick, icon, isDisabled, label, className } = {
+		next: {
+			handleClick: () => {
+				// clamp upper bound
+				const nextIdx = Math.min(pagination.page + 1, pagination.totalPages)
+				pagination.setPage(nextIdx)
+			},
+			icon: <IconChevronRight16 />,
+			label: 'Next',
+			isDisabled:
+				pagination.page === pagination.totalPages ||
+				pagination.totalPages === 0,
+			className: s.next,
+		},
+		prev: {
+			handleClick: () => {
+				// clamp lower bound (0)
+				const prevIdx = Math.max(pagination.page - 1, 0)
+				pagination.setPage(prevIdx)
+			},
+			icon: <IconChevronLeft16 />,
+			label: 'Previous',
+			isDisabled: pagination.page === 1 || pagination.totalPages === 0,
+			className: s.prev,
+		},
+	}[direction]
+
+	return (
+		<button
+			className={classNames(s.arrow, s.control, className)}
+			aria-label={label}
+			onClick={handleClick}
+			disabled={isDisabled}
+		>
+			{icon}
+			{type === 'compact' && (
+				<span className={classNames(s.label)}>{label}</span>
+			)}
+		</button>
+	)
+}
+
+const ButtonNumber = ({ page }: ButtonNumberProps) => {
+	const pagination = usePagination()
+	const handleClick = () => {
+		pagination.setPage(page)
+	}
+	return (
+		<button
+			className={classNames(s.control)}
+			type="button"
```

Comment:
💭 can we add `aria-current="page"` to the active button? This will announce the to user which page is active when focusing the active button.

![CleanShot 2023-01-12 at 09 13 51@2x](https://user-images.githubusercontent.com/825855/212089460-b5ecc526-af81-4ff9-b8c2-5aae9fb310dc.png)

Reply:
Done: https://github.com/hashicorp/dev-portal/pull/1508/commits/e779e23e279e70368d9278e36ccd5a1de5018dbf

Reply:
According to [the `aria-current` documentation](https://www.w3.org/TR/wai-aria-1.2/#aria-current), I think the `"page"` token is only valid when it's attached to an `<a>` element.

> A page token used to indicate a link within a set of pagination links, where the link is visually styled to represent the currently-displayed page.

In this component, I think we'll need to use [`aria-selected`](https://www.w3.org/TR/wai-aria-1.2/#aria-selected) (value is a boolean) instead. From the `aria-current` docs:

> Authors SHOULD NOT use the [aria-current](https://www.w3.org/TR/wai-aria-1.2/#aria-current) attribute as a substitute for [aria-selected](https://www.w3.org/TR/wai-aria-1.2/#aria-selected) in widgets where [aria-selected](https://www.w3.org/TR/wai-aria-1.2/#aria-selected) has the same meaning. For example, in a [tablist](https://www.w3.org/TR/wai-aria-1.2/#tablist), [aria-selected](https://www.w3.org/TR/wai-aria-1.2/#aria-selected) is used on a [tab](https://www.w3.org/TR/wai-aria-1.2/#tab) to indicate the currently-displayed [tabpanel](https://www.w3.org/TR/wai-aria-1.2/#tabpanel).